### ISSUE TITLE
Diff: Highlight missing \n correctly

### ISSUE THREAD

**Issue by Makman2** [coala-analyzer/coala#2006]
This is what someone gets when no \n at EOF is provided (patch from the SpaceConsistencyBear):
![tmp](https://cloud.githubusercontent.com/assets/6023916/14480942/468e1432-012e-11e6-8224-7eb3c02606ac.png)

We need to adjust the diff output so it displays the unified diff correctly (maybe using a special character, or like git does, what's best :D).

**Comment by sils1297** [comment]
see also http://bugs.python.org/issue2142

**Comment by rmariano** [comment]
As agreed with the team and @sils1297 we should handle the edge case for the diff when there are no `\n` at EOF, by adding a message displaying `\ No newline at end of file`

```diff
diff --git a/coalib/bearlib/abstractions/Linter.py b/coalib/bearlib/abstractions/Linter.py
index 9be433c..d26fc5b 100644
--- a/coalib/bearlib/abstractions/Linter.py
+++ b/coalib/bearlib/abstractions/Linter.py
@@ -692,4 +692,4 @@ def linter(executable: str,

     _prepare_options(options)

-    return partial(_create_linter, options=options)
+    return partial(_create_linter, options=options)
\ No newline at end of file
```

Therefore, the idea would be to show both lines (the `-` and `+` equally), plus adding the message at the end.

**Comment by Adrianzatreanu** [comment]
ping @rmariano

**Comment by rmariano** [comment]
pong @Adrianzatreanu 
I was looking at it. Should I add an exception in coalib (I was thinking in the Diff objecct), or maybe only when it's shown?

**Comment by Makman2** [comment]
What exception? :O

**Comment by rmariano** [comment]
By "exception" I meant a particular rule for that edge case (not a Python exception you usually raise :-), a way to distinguish the diff, and how to show it.

**Comment by Makman2** [comment]
I would say so

**Comment by sils1297** [comment]
@rmariano any progress?

**Comment by sils1297** [comment]
Unassigning @rmariano due to inactivity. If he's still working on this he can ping us later anyway and we can reassign.

**Comment by jayvdb** [comment]
I dont believe this is difficulty/low . I would put it at medium at least.

**Comment by aptrishu** [comment]
This seems to be a gitmate related issue. As it produced no error while testing it locally with no newline at the end of a file.

![spacebear](https://cloud.githubusercontent.com/assets/22702242/19883340/ce24a88e-a039-11e6-88fe-46738bb6549b.png)

**Comment by AbdealiJK** [comment]
Or maybe it has already been fixed and that is from an older version of
coala

**Comment by aptrishu** [comment]
No, though its diff is not broken but actually It misses the warning. It should show an extra part in diff. - 
`-xxxy  
\No endline of file  
+xxxy`

**Comment by jayvdb** [comment]
I am very skeptical about the approach in #2942.  Adding a copy of a stdlib module is a huge maintenance burden.

Does this only happen with `enforce_newline_at_EOF` is disabled?
Because that flag should add the missing EOL at EOF, which fixes the diff algorithm.
And if that is happening, we only need to determine the desired approach for `enforce_newline_at_EOF=False` , and IMO this could be done by having the File object always return EOLs even at EOF, and have the File object record that the EOF was missing.

**Comment by aptrishu** [comment]
The spaceConsistencyBear output
with enforce_newline_at_EOF=False
![eoltrue](https://cloud.githubusercontent.com/assets/22702242/20457342/5933f6d6-aeaf-11e6-83d2-e3119534881c.png)
with enforce_newline_at_EOF=True (default)
![eolfalse](https://cloud.githubusercontent.com/assets/22702242/20457354/73236072-aeaf-11e6-9543-199dbf4e1bc7.png)

So, spaceConsistencyBear does the check like if `enforce_newline_at_EOF=True` and newline at EOF is missing it return `No newline at  EOF` otherwise it doesn't perform that check. But this _no newline error_ thing is provided by `spaceConsistencyBear` right ? But it should be provided by DIff algo. That is why Gitmate is showing that weird diff output as it only directly uses Diff.
Moreover output of spaceConsistencyBear is not sufficient as the output after analysis should should the `\No newline at EOF` warning exactly below the affected line and that would be more understandable.

**Comment by jayvdb** [comment]
Please stop building and discuss.  What are our goals?  Find the options to achieve those goals.  Is our goal to show `\ No newline at end of file`?  I believe it is, if we want users to be able to copy & apply the patch *and* retain the missing EOL.

If you want to fix the diff algorithm, you should help at http://bugs.python.org/issue2142 .

We can already solve this bug by taking the patched `difflib.py` from http://bugs.python.org/issue2142 , and including that file as a 'backport'.  So, building another (different) patched `difflib.py` isn't helpful.

Also if you have copied any code from the patch on http://bugs.python.org/issue2142 , then that code is under a different license than `coala` and you must follow the terms of that license.  Have you done that?

The other approach is to wrap `difflib`.   I know the faults of `difflib` better than most, having been the code reviewer for most of the code at https://github.com/wikimedia/pywikibot-core/blob/master/pywikibot/diff.py , which handles hundreds of thousands of diffs per day, and difflib is full of oddities that need to be patched.  This isnt one of them.  The caller avoid this bug by appending a EOL at EOF if it is missing before calling `difflib`, and note if it was missing.  It is that simple.  It is about a 10 line wrapper to do this.  Then the wrapper can add the "missing EOL" footers if that is the goal.

If there is really a need for a replacement of `difflib` (either backport or wrapper), then it would belong in `coala_utils`, not in `coala`.

**Comment by aptrishu** [comment]
Ok so I tried to achieve the same thing without modifying the stdlib's `unified_diff()` method.
I did some preprocessing and called the unified_diff() normally, then I postprocessed the output.
I ran my code through tests and it worked :) .
Can you please take a look at my code http://pastebin.com/M1ZJag3Y .
@jayvdb @Makman2

**Comment by jayvdb** [comment]
That looks much better, and should be added to [`coala_utils`](https://gitlab.com/coala/coala-utils) with tests.

**Comment by aptrishu** [comment]
But it needs more modifications to behave exactly the same like the previous code, and it'd be much complicated.
Can't we put that previous code(modification of difflib's unified_diff method) in coala_utils ? because that code is way better and full proof. I'll put that with the licence file there. and as far as burden of handling that code is concerned, it's merely a 10 lines codes' method. wouldn't be a problem I think ?

**Comment by jayvdb** [comment]
If you think that is the better approach, do it.

**Comment by meetmangukiya** [comment]
@aptrishu is this bug fixed ?

**Comment by aptrishu** [comment]
there is a PR in gitlab, review pending..

**Comment by jayvdb** [comment]
This is waiting on https://gitlab.com/coala/coala-utils/merge_requests/26 , which I've scheduled for the next minor release.
